File: concepts/sorting/optimized_quick_sort.cpp
#include <bits/stdc++.h>
using namespace std;

#define int long long
#define endl '\n'
#define pb push_back
#define all(x) (x).begin(), (x).end()
#define yes cout << "YES" << findSmalll
#define no cout << "NO" << findSmalll

const int MOD = 1e9 + 7;
const int INF = LLONG_MAX;
void print(const vector<int> &arr)
{
    int n = arr.size();
    for (int i = 0; i < n; ++i)
        cout << "-----";
    cout << "-\n";

    for (int i = 0; i < n; ++i)
        cout << "| " << setw(2) << i << " ";
    cout << "|\n";

    for (int i = 0; i < n; ++i)
        cout << "     ";
    cout << "\n";

    for (int i = 0; i < n; ++i)
        cout << "| " << setw(2) << arr[i] << " ";
    cout << "|\n";

    for (int i = 0; i < n; ++i)
        cout << "-----";
    cout << "-\n";
}


int partition(vector<int> &arr, int low, int high)
{
    int mid = low + (high - low) / 2;
    
    swap(arr[low], arr[mid]); 

    int pivot = arr[low];
    int left = low;
    int right = high;

    while (left < right)
    {
        while (arr[left] <= pivot && left < high)
            left++;

        while (arr[right] > pivot)
            right--;

        if (left < right)
            swap(arr[left], arr[right]);
    }

    swap(arr[low], arr[right]);
    return right;
}



void quickSort(vector<int> &arr, int low, int high)
{
    if (low < high)
    {
        int p = partition(arr, low, high);
        quickSort(arr, low, p - 1);
        quickSort(arr, p + 1, high);
    }
}

int32_t main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);

    vector<int> arr = {19, 18, 16, 17, 9, 9, 7, 8, 0, 1, 2, 3, 15, 14, 12, 11, 10, 9};

    print(arr);
    quickSort(arr, 0, arr.size() - 1);
    print(arr);

    return 0;
}

File: concepts/sudo_code_practice/optimized_quick_sort.py
from template import *


def median(arr, low, high):
    mid = low + (high - low) // 2

    if arr[low] > arr[mid]:
        swap(arr, low, mid)

    if arr[low] > arr[high]:
        swap(arr, low, high)
        
    if arr[mid] > arr[high]:
        swap(arr, mid, high)

    swap(arr, mid, high)

    return arr[high]


def partition(arr, low, high):

    pivot = median(arr, low, high)

    i = low - 1

    for j in range(low, high):

        if arr[j] < pivot:
            i += 1
            swap(arr, i, j)

    swap(arr, i + 1, high)

    return i + 1


def quickSort(arr, low, high):

    if low < high:
        p = partition(arr, low, high)
        quickSort(arr, low, p - 1)
        quickSort(arr, p + 1, high)


def main():
    arr = [0, 39, 48, 87, 68, 5, 4, 3, 2, 1]
    printf(arr)
    quickSort(arr, 0, len(arr) - 1)
    printf(arr)


if __name__ == "__main__":
    main()
